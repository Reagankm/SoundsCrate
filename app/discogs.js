/* @flow */
import { secureFetch } from './lib/oauth';
import conf from './conf';
import cache from './cache';

import type {
  Folder,
  RecordsPage,
  FolderPage,
  FolderPageReleases
} from './discogsTypes';

export async function getCollectionFolder(folderId: string, pageNumber: number): Promise<RecordsPage> {
  const cachedFolder = cache.get(folderId, pageNumber);
  if (cachedFolder) {
    const { totalPages, records } = cachedFolder;
    return { totalPages, records };
  } else {
     const { totalPages, releases } = await getCollectionFolderPage(folderId, pageNumber);
     const records = releases.map(getReleaseCoverImage);
     const entry = { pageNumber, records };
     cache.update({
       folderId,
       entry,
       totalPages
     });

     return { totalPages, records };
  }
}

export async function getThumbsInFolder(folderId: string, numberOfThumbs: number = 4): Promise<*> {
  const { totalPages, releases } = await getCollectionFolderPage(folderId, 1, numberOfThumbs);
  return releases.map(getReleaseCoverImage);
}

export async function getCollectionFolders(): Promise<Array<Folder>> {
  const {
    discogs: {
      api_url: apiUrl,
      endpoints,
    }
  } = conf;
  const { username } = await getIdentity();
  const foldersCollectionUrl = endpoints.folders.replace('{username}', username);
  const foldersUrl = `${apiUrl}${foldersCollectionUrl}`;
  const foldersResponse = await secureFetch(foldersUrl);
  const { folders } = await getJson(foldersResponse);

  return folders.map(folder => ({
    id: folder.id,
    name: folder.name,
    count: folder.count,
  }));
}

async function getCollectionFolderPage(
  folderId: string,
  pageNumber: number,
  recordsPerPage?: number): Promise<FolderPage> {
  const {
    discogs: {
      api_url: apiUrl,
      endpoints,
      records_per_page,
    }
  } = conf;
  const { username } = await getIdentity();
  const perPage = recordsPerPage || records_per_page;
  const userFolderUrl = endpoints.folder.replace('{username}', username).replace('{id}', folderId.toString());
  const folderUrl = `${apiUrl}${userFolderUrl}?sort=artist&per_page=${perPage}&page=${pageNumber}`;
  const folderResponse = await secureFetch(folderUrl);
  const {
    releases,
    pagination: { pages: totalPages }
  } = await getJson(folderResponse);

  return { totalPages, releases };
}

async function getIdentity(): Promise<*> {
  const {
    discogs: {
      api_url: apiUrl,
      endpoints,
    }
  } = conf;
  const response = await secureFetch(`${apiUrl}${endpoints.identity}`);
  return await getJson(response);
}

function getJson(response: Response): Promise<*> {
  return response.json();
}

function getReleaseCoverImage(release: FolderPageReleases): string {
  return release.basic_information.cover_image;
}
