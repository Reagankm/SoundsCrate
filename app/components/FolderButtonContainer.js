/* @flow */
import React, { Component } from 'react';
import { getThumbsInFolder } from '../discogs';
import type { Folder } from '../discogsTypes';
import FolderButton from './FolderButton'

type Props = { folder: Folder };

export default class FolderButtonContainer extends Component<Props, *> {
  props: Props;
  state: { thumbnails: Array<string> } = { thumbnails: [] };

  componentDidMount() {
    this.getFolderThumbs();
  }

  render(): React$Node {
    const { folder } = this.props;
    const { thumbnails } = this.state;
    return (
      <FolderButton
        folder={folder}
        thumbnails={thumbnails}/>
    );
  }

  async getFolderThumbs(): Promise<*> {
    const { id, count } = this.props.folder;
    const thumbs = await getThumbsInFolder(id, Math.min(count, 4));
    this.setState({ thumbnails: thumbs });
  }
}
