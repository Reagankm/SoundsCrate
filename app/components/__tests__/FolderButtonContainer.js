/* @flow */
import React from 'react';
import renderer from 'react-test-renderer';
import FolderButtonContainer from '../FolderButtonContainer';
import { getThumbsInFolder } from '../../discogs';

jest.mock('../FolderButton', () => 'FolderButton');
jest.mock('../../discogs', () => ({
  __esModule: true,
  getThumbsInFolder: jest.fn()
}));

describe('app/components/FolderButtonContainer', () => {
  it('should render and call discogs.getThumbsInFolder', async () => {
    const tree = await renderer.create(
      <FolderButtonContainer
        folder={{
          id: '1',
          count: 2,
          name: 'foo'
        }} />);
    expect(getThumbsInFolder).toHaveBeenCalledWith(1, 2);
    expect(tree).toMatchSnapshot();
  });
});
