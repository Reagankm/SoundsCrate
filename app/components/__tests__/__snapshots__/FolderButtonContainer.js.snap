// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/components/FolderButtonContainer should render and call discogs.getThumbsInFolder 1`] = `
<FolderButton
  folder={
    Object {
      "count": 2,
      "id": 1,
      "name": "foo",
    }
  }
  thumbnails={undefined}
/>
`;
